### ISSUE TITLE
Cadence suddenly has issues connecting to Cassandra

### ISSUE THREAD

**Issue by slayerjain** [uber/cadence#3516]
**Describe the bug**
Cadence servers are failing to connect to Cassandra. 

**To Reproduce**
Is the issue reproducible? 
- Not sure

**Logs**
Here are some logs from the history service. 
```
2020-09-15T15:03:24.000030572Z {"level":"warn","ts":"2020-09-15T15:03:23.999Z","msg":"Failed to fetch key from dynamic config","key":"history.shardUpdateMinInterval","error":"unable to find key","logging-call-at":"config.go:61"}
2020-09-15T15:03:25.597329022Z 2020/09/15 15:03:25 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:03:27.820344435Z 2020/09/15 15:03:27 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:03:30.139261172Z 2020/09/15 15:03:30 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:03:32.564781779Z {"level":"warn","ts":"2020-09-15T15:03:32.564Z","msg":"Failed to fetch key from dynamic config","key":"history.timerProcessorCompleteTimerFailureRetryCount","error":"unable to find key","logging-call-at":"config.go:61"}
2020-09-15T15:03:41.041217459Z {"level":"warn","ts":"2020-09-15T15:03:41.041Z","msg":"Failed to fetch key from dynamic config","key":"history.timerProcessorMaxTimeShift","error":"unable to find key","logging-call-at":"config.go:61"}
2020-09-15T15:03:52.970888324Z {"level":"warn","ts":"2020-09-15T15:03:52.970Z","msg":"Failed to fetch key from dynamic config","key":"history.transferProcessorUpdateAckIntervalJitterCoefficient","error":"unable to find key","logging-call-at":"config.go:61"}
2020-09-15T15:04:02.191484311Z {"level":"warn","ts":"2020-09-15T15:04:02.191Z","msg":"Failed to fetch key from dynamic config","key":"history.timerProcessorUpdateAckInterval","error":"unable to find key","logging-call-at":"config.go:61"}
2020-09-15T15:04:15.295393747Z {"level":"warn","ts":"2020-09-15T15:04:15.295Z","msg":"Failed to fetch key from dynamic config","key":"history.timerProcessorUpdateAckIntervalJitterCoefficient","error":"unable to find key","logging-call-at":"config.go:61"}
2020-09-15T15:04:19.048284051Z 2020/09/15 15:04:19 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:04:19.420157965Z 2020/09/15 15:04:19 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:04:21.674700570Z 2020/09/15 15:04:21 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:04:23.350857848Z 2020/09/15 15:04:23 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:04:25.597122282Z 2020/09/15 15:04:25 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:04:27.820414203Z 2020/09/15 15:04:27 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:04:28.124820222Z {"level":"warn","ts":"2020-09-15T15:04:28.124Z","msg":"Failed to fetch key from dynamic config","key":"history.transferTaskBatchSize","error":"unable to find key","logging-call-at":"config.go:61"}
2020-09-15T15:04:30.139128273Z 2020/09/15 15:04:30 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:04:34.079986695Z {"level":"warn","ts":"2020-09-15T15:04:34.079Z","msg":"Failed to fetch key from dynamic config","key":"history.timerProcessorCompleteTimerInterval","error":"unable to find key","logging-call-at":"config.go:61"}
```

Here are some logs from the web container
```
nexpectedErrorError {
2020-09-15T13:50:46.836722525Z   name: 'TchannelUnexpectedError',
2020-09-15T13:50:46.836726266Z   fullType: 'tchannel.unexpected',
2020-09-15T13:50:46.836744737Z   type: 'tchannel.unexpected',
2020-09-15T13:50:46.836748442Z   message:
2020-09-15T13:50:46.836751639Z    'cadence internal error, msg: GetDomain operation failed. Error gocql: no hosts available in the pool',
2020-09-15T13:50:46.836755045Z   isErrorFrame: true,
2020-09-15T13:50:46.836757992Z   codeName: 'UnexpectedError',
2020-09-15T13:50:46.836760893Z   errorCode: 5,
2020-09-15T13:50:46.836764077Z   originalId: 5,
2020-09-15T13:50:46.836767129Z   remoteAddr: '192.168.143.191:7933' }
2020-09-15T14:10:28.421094663Z UnexpectedErrorError {
2020-09-15T14:10:28.421112662Z   name: 'TchannelUnexpectedError',
2020-09-15T14:10:28.421117014Z   fullType: 'tchannel.unexpected',
2020-09-15T14:10:28.421120307Z   type: 'tchannel.unexpected',
2020-09-15T14:10:28.421123646Z   message:
2020-09-15T14:10:28.421126954Z    'cadence internal error, msg: GetDomain operation failed. Error gocql: no hosts available in the pool',
2020-09-15T14:10:28.421130809Z   isErrorFrame: true,
2020-09-15T14:10:28.421133829Z   codeName: 'UnexpectedError',
2020-09-15T14:10:28.421151456Z   errorCode: 5,
2020-09-15T14:10:28.421154911Z   originalId: 5,
2020-09-15T14:10:28.421158333Z   remoteAddr: '192.168.143.191:7933' }
```

here are some logs from the cdence worker
```
2020-09-15T15:09:04.029288594Z 2020/09/15 15:09:04 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:09:04.688595691Z 2020/09/15 15:09:04 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:09:06.977125394Z 2020/09/15 15:09:06 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:09:07.724804936Z 2020/09/15 15:09:07 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:09:10.087949722Z 2020/09/15 15:09:10 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:09:30.11345922Z {"level":"warn","ts":"2020-09-15T15:09:30.113Z","msg":"Failed to fetch key from dynamic config","key":"system.enableGlobalDomain","error":"unable to find key","logging-call-at":"config.go:61"}
2020-09-15T15:10:03.678872177Z 2020/09/15 15:10:03 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:10:05.958349923Z 2020/09/15 15:10:05 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:10:06.580876582Z 2020/09/15 15:10:06 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:10:06.977223511Z 2020/09/15 15:10:06 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:10:07.726221878Z 2020/09/15 15:10:07 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:10:10.08800695Z 2020/09/15 15:10:10 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:11:02.192633546Z 2020/09/15 15:11:02 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:11:04.0292871Z 2020/09/15 15:11:04 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:11:04.688873449Z 2020/09/15 15:11:04 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:11:08.314044113Z 2020/09/15 15:11:08 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:11:08.358146752Z 2020/09/15 15:11:08 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
2020-09-15T15:11:10.088037544Z 2020/09/15 15:11:10 error: failed to connect to 192.168.106.144:9042 due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
```

**Additional context**
We're using Cassandra 3.11.7 3 node cluster.

**Comment by longquanzheng** [comment]
``` 
due to error: Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)
```

What is your Cassandra version? seems like in-compatible with the gocql.

**Comment by longquanzheng** [comment]
I submitted an feature request here: https://github.com/uber/cadence/issues/3536

**Comment by ttreptow** [comment]
I ran into the same problem in Temporal with the same persistence setup as @slayerjain  - Cassandra 3.11.7 three node cluster (created by cass-operator in kubernetes in my case). It happened on a temporal cluster that had been working fine. It started immediately after a cassandra node restarted. The problem initially only occurred with that node and temporal continued to function even though the services were printing out the error. Eventually (after at least a week) temporal stopped functioning altogether.

I dug into the Cassandra code that is throwing the error and cannot see how it can logically print out the error message "Invalid or unsupported protocol version (4); supported versions are (3/v3, 4/v4, 5/v5-beta)". This issue is the only reference to this exact message on the entire internet.

I managed to fix(?) the issue by blowing away the cassandra cluster (including storage) and recreating it. This may have been more drastic than necessary but it was ok for me since it was a test cluster. I didn't try expanding the cassandra cluster or deleting the disks one by one for example. 

I'm not sure that there's anything that the cadence project can do to fix this (assuming it's even a cadence problem) as there isn't much information to go on, but I wanted to document what I saw for the next frustrated dev looking for answers.
